Changes for internal reporting

kpthatsme opened the pull request:
@yakkomajuri I made a few changes to how the plugin works:

We send 2 events:
1. An Upcoming Invoice event for all pending invoices - tying quantity, amount, product to it
2. An Upcoming Invoices (Aggregated) event summing the total for each product type

This let's us build a pretty good dashboard w/
- Subscribers (growth + aggs, by plan)
- Upcoming Invoices (Growth + Agg, by plan)
- Top 10% of subscribers by quantity + total

_(plugin is running locally and screenshot has sensitive data so will share the dash via slack)_

Some things I removed:
- Caching – we don't need to cache any users since there isn't an issue of double runs. Also caching on invoice creation time won't work for metered billing.
- Auto-adding all the stripe props to the user. I think adding all the props introduces more confusion and data cleanliness issues, i'd rather have a use case in mind before dumping all of them in there. I'll make it easier to onboard people through this too.
- The application of threshold based notifications – I think this new top 10% chart should handle the initial ask from @jamesefhawkins. I do think there's still room for something here. Perhaps more it can be its own thing, i.e. an alert in addition to receiving billing updates, an additional event that fires or something. There could be multiple kinds of alerts that this plugin supports. The broad solution in our product for all these kinds of problems is going to be monitoring / alerting based on specific events and properties so maybe we should look to route through there in the long term. I'd opt for keeping the plugin simpler at first.
- Commented out the Subscriber Created / Updated events – similar to above I think if we want these maybe it should be it's own event. It wasn't clear to me the need for a separate event firing here, or if it was a very useful event now that we can pull subscribers another way. qq: What's the life expectancy of `storage`? 


I want to use this internally asap for PH news and if it goes smoothly, I think this could be a good 'learn growth through the product' kind of experience.

Commit:
initial commit - plugin for rev dash

Commit:
update pagination + add a comment

Commit:
update tests

Commit:
add return back

yakkomajuri commented on index.js:
```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
         }
 
-        const customerRecordExists = await storage.get(customer.id)
+        // Note: I think if we want to automatically update users we should just update properties instead of firing events
```
Problem is we don't have access to `posthog.people.set` inside plugins yet, so firing an event is actually just a proxy for setting the props.

The alternative is to use the PostHog API, but that adds complexity as it needs a PostHog API key.

kpthatsme replied:
makes sense – thanks!

yakkomajuri commented on index.js:
```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
-        const customerRecordExists = await storage.get(customer.id)
+        // Note: I think if we want to automatically update users we should just update properties instead of firing events
+        // I also think we should perhaps restrict the properties we send so it doesn't feel overwhelming to
+        // see a lot of properties you may not understand.
```
Yeah that was definitely a lot - the goal was to give people all the info because every company will have a different setup. I ultimately intended to either:
a) Build some schema enforcing here
b) Steer people towards using the schema enforcer plugin

kpthatsme replied:
Yea, I definitely see the value in getting a lot of data in upfront and then helping make it useful downstream. I think the longer term solution here is going to be something in product (ie a taxonomy product). I like the in-between idea of only setting basic props for now, so going to go with that for now.

yakkomajuri commented on index.js:
```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
+        //     break
+        // }
+        //
+        // posthog.capture(customerRecordExists ? 'update_stripe_customer' : 'new_stripe_customer', {
```
Maybe updating customer was too much (it wasn't on by default either) but I think it'd be useful to get new users into PostHog. We can maybe just use the `basicProperties` for now?

yakkomajuri commented on index.js:
```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
+        //     ...properties,
+        //     $set: basicProperties
+        // })
+    }
```
Used this section for some questions, but otherwise let's remove the comments - no need to have this lying around

yakkomajuri reviewed (commented):
Great stuff!

Code-wise looks good to me.

As for some of the open Qs:

- I don't believe there's a limit on storage currently: https://github.com/PostHog/plugin-server/blob/master/src/vm/extensions/storage.ts (maybe there should be?)
- The invoice threshold mechanism was what I built to allow for the alerting - it wasn't made to calculate revenue (which this does!). I created an action with a webhook to Slack from the `upcoming_invoice` event. I think that might still be useful (as a setting you need to toggle on). The cache there was used to make sure an upcoming invoice was only registered once per cycle according to the specified params.

And will leave this open while we discuss these questions but otherwise happy to merge and start using it! I won't put it on the repository yet so this will be for our internal use before we put it out to the public

yakkomajuri commented on index.js:
```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
+        const props = {
+            amount: parseFloat(billingSum.toFixed(2)),
+            product: product,
+            distinct_id: 'posthog-stripe-plugin'
```
This isn't necessary - events with no `distinct_id` set will automatically be associated with the "plugin user"

Commit:
re-enable customer identification + add an event for billing alerts

kpthatsme commented:
> Great stuff!
> 
> Code-wise looks good to me.
> 
> As for some of the open Qs:
> 
> * I don't believe there's a limit on storage currently: https://github.com/PostHog/plugin-server/blob/master/src/vm/extensions/storage.ts (maybe there should be?)
> * The invoice threshold mechanism was what I built to allow for the alerting - it wasn't made to calculate revenue (which this does!). I created an action with a webhook to Slack from the `upcoming_invoice` event. I think that might still be useful (as a setting you need to toggle on). The cache there was used to make sure an upcoming invoice was only registered once per cycle according to the specified params.
> 
> And will leave this open while we discuss these questions but otherwise happy to merge and start using it! I won't put it on the repository yet so this will be for our internal use before we put it out to the public

@yakkomajuri thanks for the review!

A couple of updates

1. I added Alert thresholds back in – these will be sent as a _second_ event if the invoice amount is > threshold
2. I added the registration / updating of stripe customers back in– sending the basic props + product name instead of all sub info.
3. Removed the distinct_id from the agg call

Commit:
update tests
